### PR TITLE
Feature/issue 29 pattern match input state

### PR DESCRIPTION
**Pull request by dzimmanck**
In this pull request:

1.  Added support for pattern matching on input states
2.  Added support for a "wildcard" input state with "_"
3.  Re-factored parser.rs into 7 separate concise files for ease of readability/maintainability.
4.  Started adding unit tests (input_state.rs)

### CODE REVIEW

File: macros/src/parser/input_state.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    fn parse(input: parse::ParseStream) -> syn::Result<Self> {
+        // Check for starting state definition
+        let start = input.parse::<Token![*]>().is_ok();
+
+        // check to see if this is a wildcard state, which is denoted with "underscore"
+        let underscore = input.parse::<Token![_]>();
+        let wildcard = underscore.is_ok();
+
+        // Input State
+        let ident: Ident = if let Ok(underscore) = underscore {
```
Comment (korken89) [inline macros/src/parser/input_state.rs:21]: What happens if both start and wildcard are used together? It looks like it's allowed.

```
*_ + ...
```

**Review by korken89 — CHANGES_REQUESTED** [review]
Looks good, just a small thing I found.

Also, update the README - it has a contributors section. Add yourself there. :)

File: macros/src/parser/input_state.rs (hunk 1 above)
Comment (dzimmanck) [reply macros/src/parser/input_state.rs:29]: Good catch.  This should not be allowed.  I added code to panic if this occurs with an appropriate message, as well as an additional unit test to cover.  I also updated the ReadMe.

**Review by korken89 — APPROVED** [review]
bors merge

**Pull request merged**